### PR TITLE
MG-271: Implement standard must-gather directory naming convention

### PR DESCRIPTION
**Pull request by neha037**
## Summary

Implements standard must-gather directory naming convention to align with `oc adm must-gather`, ensuring consistency across all must-gather collection methods. Also replaces runtime `SubPathExpr`/`$(POD_NAME)` with static `SubPath` using the generated directory name for PVC storage isolation.

### Implementation
- **New**: `pkg/mustgatherutil/util.go` — Directory name generation utility
  - `GenerateMustGatherDirectoryName()`: Generates directory names following the `oc adm must-gather` convention
  - `getClusterIDSuffix()`: Retrieves last 12 characters of cluster ID from ClusterVersion resource
  - `generateRandomSuffix()`: Generates a random numeric suffix (at least 6 digits, zero-padded) for uniqueness
- **Modified**: `controllers/mustgather/mustgather_controller.go`
  - Reordered reconcile flow to look up existing Job before building template (avoids unnecessary ClusterVersion API calls on re-reconcile)
  - Unconditionally generates directory name via `mustgatherutil.GenerateMustGatherDirectoryName()` inside `getJobFromInstance()`
- **Modified**: `controllers/mustgather/template.go`
  - Replaced `outputSubPathExpr()` with `outputSubPath()` — uses static `SubPath` with generated directory name instead of runtime `SubPathExpr`/`$(POD_NAME)`
  - Removed `podNameEnvVars()` function and `POD_NAME` env var injection (no longer needed)
  - `getGatherContainer()` and `getUploadContainer()` accept `directoryName` for PVC `SubPath` isolation
  - `getUploadContainer()` injects `FILENAME_PREFIX` env var with the directory name
- **Modified**: `controllers/mustgather/constant.go` — Removed `podNameEnvVar` constant
- **Modified**: `build/bin/upload` — Uses `FILENAME_PREFIX` as the archive filename when set; falls back to `must-gather-<timestamp>.tar.gz` when unset

### Tests
- **New**: `pkg/mustgatherutil/util_test.go`
  - `TestGenerateMustGatherDirectoryName` — 5 sub-tests: format with full UUID, exactly 12 chars, 13 chars (boundary), ClusterVersion not found (fallback), empty cluster ID (fallback)
  - `TestGenerateMustGatherDirectoryName_UTCConversion` — Verifies non-UTC timestamps are converted to UTC
  - `TestGetClusterIDSuffix_WrongClusterVersionName` — Fallback when ClusterVersion name doesn't match
- **Modified**: `controllers/mustgather/mustgather_controller_test.go`
  - `reconcile_job_created_with_FILENAME_PREFIX_env_var` — Verifies FILENAME_PREFIX with correct format and cluster ID suffix
  - `reconcile_existing_job_does_not_regenerate_directory_name` — Idempotent re-reconcile does not recreate Job
  - `reconcile_job_created_with_PVC_uses_generated_subpath` — PVC storage generates SubPath (not SubPathExpr), no `$(POD_NAME)`
- **Modified**: `controllers/mustgather/template_test.go`
  - `Test_getJobTemplate_FilenamePrefix` — FILENAME_PREFIX present when directoryName is set, absent when empty
  - `Test_outputSubPath` — Edge cases: nil storage, empty/whitespace/slash subPath, empty directoryName
  - Updated PVC assertions to check `SubPath` (not `SubPathExpr`)
- **Modified**: `test/e2e/must_gather_operator_test.go`
  - "Directory Naming Convention Tests": FILENAME_PREFIX format validation on upload container, uniqueness across concurrent CRs
  - Updated PVC storage tests to assert `SubPath` prefix matches `must-gather.local.` and `SubPathExpr` is empty

### With Cluster ID (default)
```
must-gather.local.<cluster-id-suffix>.<timestamp>.<random>
```
Example: `must-gather.local.fa96a0bf69fd.20260617T143025Z.4394764343467089039`

### Without Cluster ID (fallback)
```
must-gather.local.<timestamp>.<random>
```
Example: `must-gather.local.20260617T143025Z.042315`

Where:
- **cluster-id-suffix**: Last 12 characters of cluster ID from `clusterversion.config.openshift.io/version`
- **timestamp**: UTC timestamp in format `YYYYMMDDTHHMMSSZ`
- **random**: Random numeric suffix (at least 6 digits, zero-padded via `fmt.Sprintf("%06d", rand.Int63())`)

## RBAC

RBAC for ClusterVersion access is already present:
```yaml
//+kubebuilder:rbac:groups=config.openshift.io,resources=clusterversions,verbs=get;list;watch
```

### Unit Tests
```bash
go test ./pkg/mustgatherutil/... ./controllers/mustgather/... -v
```

### E2E Tests
```bash
go test -tags e2e -timeout 30m ./test/e2e/... -v -ginkgo.v -ginkgo.focus="Directory Naming Convention"
```

## Verification

```bash
go build ./...        # ✅ Builds successfully
go test ./...         # ✅ All tests pass
```

## Additional Notes

- **No API Changes**: Implementation is transparent to users — no changes to MustGather CRD
- **PVC Improvement**: PVC subdirectory names are now human-readable (`must-gather.local.<cluster>.<timestamp>.<random>`) instead of opaque Kubernetes pod names
- **Fallback Behavior**: Gracefully handles scenarios where cluster ID is unavailable (permissions, missing resource) — the name simply omits the cluster ID segment

## Implements

Jira: [MG-271](https://redhat.atlassian.net/browse/MG-271) — Change must-gather directory name in must-gather-operator

### Acceptance Criteria
- ✅ Directory naming convention is consistent with `oc adm must-gather`
- ✅ PVC storage uses static SubPath with generated directory name (not SubPathExpr/POD_NAME)
- ✅ Relevant unit tests, controller tests, and E2E tests added

[MG-271]: https://redhat.atlassian.net/browse/MG-271?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: controllers/mustgather/directory_naming.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+// generateMustGatherDirectoryName generates a directory name following the same convention as oc adm must-gather.
+// Format: must-gather.local.<cluster-id-suffix>.<timestamp>.<random>
+// If cluster ID is unavailable: must-gather.local.<timestamp>.<random>
+func generateMustGatherDirectoryName(ctx context.Context, c client.Client, now time.Time) string {
+	parts := []string{"must-gather.local"}
+
+	clusterIDSuffix, err := getClusterIDSuffix(ctx, c)
+	if err != nil {
+		log.V(2).Info("Unable to retrieve cluster ID, continuing without it", "error", err)
```
[the remark this answers is not in the document]
Comment (neha037) [reply controllers/mustgather/directory_naming.go]: Skipping this finding. The error here comes from getClusterIDSuffix which returns one of three messages: "clusterversion resource not found", "insufficient permissions to read clusterversion", or "failed to get clusterversion". These are generic API errors about the
  ClusterVersion resource — they don't contain customer metadata. The V(2) log level means this only appears in verbose debug logging, and the error context is essential for operators diagnosing why the cluster ID is missing from the directory name. Removing it would
  make debugging harder with no security benefit.

File: controllers/mustgather/mustgather_controller.go [hunk 2]
```diff
@@ -176,31 +177,31 @@ func (r *MustGatherReconciler) Reconcile(ctx context.Context, request reconcile.
 		}
 	}
```
Comment (neha037) [inline controllers/mustgather/mustgather_controller.go:180]: the restructuring is done because

  Old code (master) — built the job template before checking if the job already exists:
  job := getJobFromInstance(...)     // always runs — API calls, image lookup, etc.
  job1 := client.Get(job.Name, ...) // then check if it exists
  if not found → create

  this branch — checks first, only builds if needed:
  existingJob := client.Get(instance.Name, ...) // check first
  if not found → job := getJobFromInstance(...)  // only build when we need to create

getJobFromInstance now calls generateMustGatherDirectoryName, which makes an API call to fetch the ClusterVersion. On the old code path, that API call would happen on every reconcile — even when the job already exists and is running. With restructuring, the directory name generation (and the ClusterVersion lookup) only happens once, when actually creating the job.

The rename from job1 to existingJob is just a clarity

**Comment by neha037** [comment]
@coderabbitai review

**Comment by neha037** [comment]
@coderabbitai summary

File: pkg/mustgatherutil/util.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	id := string(clusterVersion.Spec.ClusterID)
+	if len(id) > clusterIDSuffixLength {
+		return id[len(id)-clusterIDSuffixLength:]
+	}
+	return id
+}
+
+func generateRandomSuffix() string {
```
Comment (praveencodes) [inline pkg/mustgatherutil/util.go:73]: @neha037 This function doesn't generate 6 digit unique number.
https://go.dev/play/p/T4j2Ft86_T9

Comment (praveencodes) [reply pkg/mustgatherutil/util.go:73]: Pls check this - https://go.dev/play/p/PGl9RrwSQ0g

Comment (neha037) [reply pkg/mustgatherutil/util.go:73]: hey @praveencodes , 6 digit is not required according to https://redhat.atlassian.net/browse/RFE-5434 . Don'r know where where the description picked it up. Will update it.

Comment (praveencodes) [reply pkg/mustgatherutil/util.go:73]: Ok. Thanks for the update.
